From aea2f7dcd031b0eaecafca7ecc07b216ab9ccf09 Mon Sep 17 00:00:00 2001
From: Geetha sowjanya <gakula@marvell.com>
Date: Wed, 3 Feb 2021 12:00:16 +0530
Subject: [PATCH 1239/1921] octeontx2-af: TIM: Set conditional clock always on

Due to hardware bug TIM expire counter updates incorrectly
because to wrong clock gating.
This patch adds workaround for this issue by setting
FORCE_CSCLK_ENA=1

Change-Id: I6c65d53a20f20c4799685e1a16580f7ce17ed367
Signed-off-by: Geetha sowjanya <gakula@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/45231
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/af/rvu.h       |  1 +
 drivers/net/ethernet/marvell/octeontx2/af/rvu_fixes.c | 11 +++++++++++
 drivers/net/ethernet/marvell/octeontx2/af/rvu_tim.c   |  3 +++
 3 files changed, 15 insertions(+)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu.h b/drivers/net/ethernet/marvell/octeontx2/af/rvu.h
index d2dd4c31e6be..005b75662b9a 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu.h
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu.h
@@ -860,6 +860,7 @@ bool is_parse_nibble_config_valid(struct rvu *rvu,
 				  struct npc_mcam_kex *mcam_kex);
 int rvu_npc_set_parse_mode(struct rvu *rvu, u16 pcifunc, u64 mode, u8 dir,
 			   u64 pkind);
+void rvu_tim_hw_fixes(struct rvu *rvu, int blkaddr);
 
 /* CN10K RVU */
 int rvu_set_channels_base(struct rvu *rvu);
diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu_fixes.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu_fixes.c
index 5bf742341fd8..a6c7baca4e7a 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu_fixes.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu_fixes.c
@@ -1002,3 +1002,14 @@ void rvu_smqvf_xmit(struct rvu *rvu)
 		otx2smqvf_xmit();
 	}
 }
+
+void rvu_tim_hw_fixes(struct rvu *rvu, int blkaddr)
+{
+	u64 cfg;
+	/* Due wrong clock gating, TIM expire counter is updated wrongly.
+	 * Workaround is to enable force clock (FORCE_CSCLK_ENA = 1).
+	 */
+	cfg = rvu_read64(rvu, blkaddr, TIM_AF_FLAGS_REG);
+	cfg |= BIT_ULL(1);
+	rvu_write64(rvu, blkaddr, TIM_AF_FLAGS_REG, cfg);
+}
diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu_tim.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu_tim.c
index edea6c34bf68..18cd9585512f 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu_tim.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu_tim.c
@@ -355,5 +355,8 @@ int rvu_tim_init(struct rvu *rvu)
 		 BIT_ULL(0); /* ENA_TIM */
 	rvu_write64(rvu, blkaddr, TIM_AF_FLAGS_REG, regval);
 
+	if(is_rvu_otx2(rvu))
+		rvu_tim_hw_fixes(rvu, blkaddr);
+
 	return 0;
 }
-- 
2.31.1

